File: 0042-trapping-rain-water/0042-trapping-rain-water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        
        if len(height) <= 2:    # safe coding
            return 0

        ans = 0

        l = 1
        r = len(height)-1

        lmax = height[0]
        rmax = height[-1]

        while l <= r:

            if height[l] > lmax:
                lmax = height[l]
            if height[r] > rmax:
                rmax = height[r]
        

            if lmax <= rmax:
                ans += lmax - height[l]
                l += 1
            else:
                ans += rmax-height[r]
                r -= 1
        
        return ans

File: Top 100 Liked/Two Pointers/0042-trapping-rain-water/0042-trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        if (height.size() <= 2) return 0;

        long long answer = 0;
        int n = height.size();

        int l = 0;        
        int r = n - 1;   

        int lmax = height[0];
        int rmax = height[n-1];

        while (l < r)     
        {
            if (lmax <= rmax)
            {
                l++;
                if (height[l] > lmax) {
                    lmax = height[l];
                } else {
                    answer += lmax - height[l];
                }
            }
            else
            {
                r--;
                if (height[r] > rmax) {
                    rmax = height[r];
                } else {
                    answer += rmax - height[r];
                }
            }
        }

        return answer;
    }
};